### PR TITLE
Optimize secret manager to refresh secrets from apiserver cache

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #19188

@liggitt

### CODE REVIEW

File: pkg/kubelet/secret/secret_manager.go [hunk 1]
```diff
@@ -169,7 +169,15 @@ func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
 	data.Lock()
 	defer data.Unlock()
 	if data.err != nil || !s.clock.Now().Before(data.lastUpdateTime.Add(s.ttl)) {
-		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+		opts := metav1.GetOptions{}
+		if data.secret != nil && data.err == nil {
+			// This is just a periodic refresh.
+			// In this case, we just try to get data from apiserver cache to reduce the load
+			// on etcd (the cache is eventuallly consistent).
+			// We should always hit etcd if the update was explicitly requested.
```
Comment (liggitt) [inline pkg/kubelet/secret/secret_manager.go]: the "// We should always hit etcd if the update was explicitly requested." comment is confusing here, since it's describing the other case, not the one we're handling here

File: pkg/kubelet/secret/secret_manager.go [hunk 2]
```diff
@@ -169,7 +169,15 @@ func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
 	data.Lock()
 	defer data.Unlock()
 	if data.err != nil || !s.clock.Now().Before(data.lastUpdateTime.Add(s.ttl)) {
-		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+		opts := metav1.GetOptions{}
+		if data.secret != nil && data.err == nil {
+			// This is just a periodic refresh.
```
Comment (liggitt) [inline pkg/kubelet/secret/secret_manager.go]: "This is just a periodic refresh of a secret we successfully fetched previously."

**Comment by liggitt** [comment]
a testcase would be good

File: pkg/kubelet/secret/secret_manager.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: done

File: pkg/kubelet/secret/secret_manager.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: removed

**Review by wojtek-t — COMMENTED** [review]
> a testcase would be good

@liggitt - I agree, though currently fakeClient doesn't have possibility to check its options.

I will create another PR to fix the fake client and make it possibly, but if possible wouldn't like to block this PR on that.

File: pkg/kubelet/secret/secret_manager.go [hunk 3]
```diff
@@ -169,7 +169,14 @@ func (s *secretStore) Get(namespace, name string) (*v1.Secret, error) {
 	data.Lock()
 	defer data.Unlock()
 	if data.err != nil || !s.clock.Now().Before(data.lastUpdateTime.Add(s.ttl)) {
-		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+		opts := metav1.GetOptions{}
+		if data.secret != nil && data.err == nil {
+			// This is just a periodic refresh of a secret we successfully fetched previously.
+			// In this case, we just try to get data from apiserver cache to reduce the load
+			// on etcd (the cache is eventuallly consistent).
+			opts.ResourceVersion = "0"
```
Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: Is the optimization really necessary? I found it hard to follow when we have scattered apiserver code lying in the kubelet.

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: Unfortunately, it changes a lot. With 5k nodes, etcd is becoming a bottleneck, so anything we can do to offload it is useful.

Why do you thing is hard to follow? This is just one additional if (that is pretty simple) in a single place in the code. Or maybe asking it differently: any suggestions how we can make it easier?

Comment (xiang90) [inline pkg/kubelet/secret/secret_manager.go]: > With 5k nodes, etcd is becoming a bottleneck, so anything we can do to offload it is useful.

hmm... based on our experiments, at least for kube-mark, kubernetes only write to etcd ~1000op/second (a 5 nodes etcd with standard4 + PD can support ~10000 writes easily). etcd is far from becoming a bottleneck if configured correctly.

Comment (xiang90) [inline pkg/kubelet/secret/secret_manager.go]: @wojtek-t 

To be more clear, I am not saying we should not offload. I want to understand why etcd is becoming a bottleneck. Then we can try to solve it from the bottom layer if possible.

Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: I think this is something that should be dealt with in the client library. The callers should not need to know what "setting resource version = 0" means in this context.

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: @xiang90 - sorry, I wasn't very clear (and shouldn't say that it's purely etcd issue). Basically the problem is that if we are serving from etcd:
- we need to deserialize the data first (we don't do it if we serve from apiserver cache)
- this increases latency of processing request, which in turn means that we can in total process less requests, due to "max-inflight-requests" limit.
So yes - this isn't the etcd issue itself.

@yujuhong - I can work on improving the ux for it, but this isn't obvious how to do it. Basically, it's the user who need to decide whether he can or not tolerate that returned data may be slightly returned for every request. So it is the user that needs to set some bit on the request.
Though, i agree that doing it by setting 'resourceVersion=0' is not very intuitive.

However, can we merge this PR as is and work on improving UX later?

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: Also to prove my words from above, I'm seeing a bunch of logs like this in our kubemark runs:
```
I0125 13:47:20.199420       7 wrap.go:75] GET /api/v1/namespaces/e2e-tests-density-30-32-lxszp/secrets/default-token-9zsd0: (2.716521925s) 200 [[kubemark/v0.0.0 (linux/amd64) kubernetes/$Format] 104.197.210.74:41632]
```

2.7s for processing a simple GET operation is quite a lot...

Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: @wojtek-t even a simple utility function to generate or modify a `metav1.GetOptions{}` object is strictly better IMO.

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: @yujuhong will you be fine with something like:
```
func GetFromApiserverCache(opts *metav1.GetOptions) {
  opts.ResourceVersion = "0"
}
```
and changing this code to call this function (as a temporary solution)?

Comment (xiang90) [inline pkg/kubelet/secret/secret_manager.go]: @wojtek-t Thanks for the explanation. I agree the cache efficiency matters.

Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: @wojtek-t SGTM.

One question. With the optimization like this, do we still need to bump the TTL?

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: @yujuhong - thanks. Will change that tomorrow.

Regarding TTL - I'm afraid this optimization won't be enough. Though I hope that bump might be lower with that.

Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: Serving from the cache should be much faster.  It'd be good to profile the performance changes after this gets in, before we attempt to bump the TTL.

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: @yujuhong - sure, we should definitely merge this one and see the results of this improvment before the other PR.

**Comment by wojtek-t** [comment]
@yujuhong - PTAL

File: pkg/kubelet/secret/secret_manager.go [hunk 4]
```diff
@@ -145,6 +145,12 @@ func (s *secretStore) Delete(namespace, name string) {
 	}
 }
 
+// fromApiserverCache modifies <opts> so that the GET request will
```
Comment (yujuhong) [inline pkg/kubelet/secret/secret_manager.go]: How about moving this to the util package? This can be used for [getting the node object](https://github.com/kubernetes/kubernetes/blob/a507777de86f9a64ac0a1ff1ed98f7e1e23f8f4d/pkg/kubelet/kubelet_node_status.go#L352) too :-)

Other than that the rest of the PR LGTM

Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: Updated. PTAL
